### ISSUE TITLE
sysrepocfg --get doesnt return the container for the yang model it reads

### ISSUE THREAD

**Issue by sankettandulwadkar** [sysrepo/sysrepo#850]
I am trying to read a container of a yang model I added using sysrepocfg, but it doesn't output anything -
```
sanket@sanket-HP-Compaq-8200-Elite-CMT-PC:~$ sysrepocfg --get /ona100:bridge/controllers -l 3
[INF] (sr_connect:445) Connected to daemon Sysrepo Engine at socket=/tmp/sysrepod.sock
sanket@sanket-HP-Compaq-8200-Elite-CMT-PC:~$ sysrepocfg --get /ona100:bridge/controllers -l 4
[DBG] (sr_log_stderr:141) Setting log level for stderr logs to 4.
[DBG] (sr_connect:390) Connecting to Sysrepo Engine.
[DBG] (cl_socket_connect:376) Connecting to socket=/tmp/sysrepod.sock
[INF] (sr_connect:445) Connected to daemon Sysrepo Engine at socket=/tmp/sysrepod.sock
[DBG] (cl_version_verify:429) Sending version-verify request.
[DBG] (cl_version_verify:440) version-verify request sent, waiting for response.
[DBG] (cl_version_verify:452) version-verify response received, processing.
[DBG] (cl_request_process:492) Sending session-start request.
[DBG] (cl_request_process:515) session-start request sent, waiting for response.
[DBG] (cl_request_process:538) session-start response received, processing.
[DBG] (cl_request_process:492) Sending check-enabled-running request.
[DBG] (cl_request_process:515) check-enabled-running request sent, waiting for response.
[DBG] (cl_request_process:538) check-enabled-running response received, processing.
[DBG] (srcfg_load_module_schema:168) Loading module schema: '/home/sanket/sysrepo/build/repository/yang/ona100@2015-11-07.yang'.
[DBG] (cl_request_process:492) Sending get-items request.
[DBG] (cl_request_process:515) get-items request sent, waiting for response.
[DBG] (cl_request_process:538) get-items response received, processing.
[DBG] (cl_request_process:492) Sending get-items request.
[DBG] (cl_request_process:515) get-items request sent, waiting for response.
[DBG] (cl_request_process:538) get-items response received, processing.
[DBG] (sr_get_item_next:1105) All items has been read for xpath '/ona100:bridge/controllers'
[DBG] (cl_request_process:492) Sending session-stop request.
[DBG] (cl_request_process:515) session-stop request sent, waiting for response.
[DBG] (cl_request_process:538) session-stop response received, processing.
```
It states all items for the xpath are read, but nothing is displayed. How do I display it?

**Comment by fabioantonini** [comment]
Hi Sanket

did you try with

# sysrepocfg --get /ona100:bridge/controllers*/** -l 4


Regards
fabio

**Comment by michalvasko** [comment]
Hi Sanket,
what Fabio probably meant was `/ona100:bridge/controllers//.`, which will get you the `controllers` container and all of its descendants. However, there must be some relevant data for them to be displayed, You can see all the data of a model using `--export`, but I do not know the name of the model, so I cannot give you the exact command.

Regards,
Michal

**Comment by sankettandulwadkar** [comment]
Hi Michal and Fabio,

I tried the command, and it still returns nothing, although the command is
successfully executed.

sanket@sanket-HP-Compaq-8200-Elite-CMT-PC:~$ sysrepocfg --get
/ona100:bridge/controllers//. -l 4
[DBG] (sr_log_stderr:141) Setting log level for stderr logs to 4.
[DBG] (sr_connect:390) Connecting to Sysrepo Engine.
[DBG] (cl_socket_connect:376) Connecting to socket=/tmp/sysrepod.sock
[INF] (sr_connect:445) Connected to daemon Sysrepo Engine at
socket=/tmp/sysrepod.sock
[DBG] (cl_version_verify:429) Sending version-verify request.
[DBG] (cl_version_verify:440) version-verify request sent, waiting for
response.
[DBG] (cl_version_verify:452) version-verify response received, processing.
[DBG] (cl_request_process:492) Sending session-start request.
[DBG] (cl_request_process:515) session-start request sent, waiting for
response.
[DBG] (cl_request_process:538) session-start response received, processing.
[DBG] (cl_request_process:492) Sending check-enabled-running request.
[DBG] (cl_request_process:515) check-enabled-running request sent, waiting
for response.
[DBG] (cl_request_process:538) check-enabled-running response received,
processing.
[DBG] (srcfg_load_module_schema:168) Loading module schema:
'/home/sanket/sysrepo/build/repository/yang/ona100@2015-11-07.yang'.
[DBG] (cl_request_process:492) Sending get-items request.
[DBG] (cl_request_process:515) get-items request sent, waiting for response.
[DBG] (cl_request_process:538) get-items response received, processing.
[DBG] (cl_request_process:492) Sending get-items request.
[DBG] (cl_request_process:515) get-items request sent, waiting for response.
[DBG] (cl_request_process:538) get-items response received, processing.
[DBG] (sr_get_item_next:1105) All items has been read for xpath
'/ona100:bridge/controllers//.'
[DBG] (cl_request_process:492) Sending session-stop request.
[DBG] (cl_request_process:515) session-stop request sent, waiting for
response.
[DBG] (cl_request_process:538) session-stop response received, processing.


I also tried the export command and my backup.json file seems to be empty.
The model name is "ona100.yang"

sanket@sanket-HP-Compaq-8200-Elite-CMT-PC:~$ sysrepocfg
--export=/home/sanket/backup.json --format=json --datastore=startup ona100
The configuration was successfully exported.
sanket@sanket-HP-Compaq-8200-Elite-CMT-PC:~$ cat /home/sanket/backup.json

Basically, my goal is to read different leafs in a container/read entire
container. Suppose controller container has IP, username and password
paramaters, my get operation should display the current values of these
parameters and blank if nothing is present/set there. ie {IP:
"",username:"",password:""}

Is there anyway to display the entire container or leaf through CLI?


--Sanket

**Comment by michalvasko** [comment]
Hi Sanket,
all this behavior seems correct to me. For one, you are doing `--get` without `--datastore`, so the default is `running`. I presume it is empty just like `startup`, which you tried to export. So I am not sure what is it you are trying to achieve. Yes, you can display entire whatever using CLI, but there must be some data to display, which is likely not your case.

Regards,
Michal

**Comment by sankettandulwadkar** [comment]
I dont think that is correct. Reasons for that -

1) When i start the application_changes_example example, it says "reading
startup config" and shows me all the containers in ona100, then says
"startupconfig applied as running."
2) If the datastores are both empty, why does the log say "all items has
been read for xpath '/ona100:bridge/controllers//." It should say nothing
to read or something. Isn't that logical?

I did add --datastore to my commands and I see no difference.
Assuming my both datastore are empty, which could very well be true. How do
I add/verify that the added yang file is now in the datastore, whether it
be startup or running. I thought the sysrepoctl --intsall step does the
adding to datastore part. Apparently I was wrong.

**Comment by fabioantonini** [comment]
Hi Sanket

please can show the output of application_changes_example when it 
starts? What you see could be only the default values defined in the 
yang file.

How did you import your data in the datastore before running 
application_changes_example?

I want to remark that the sysrepoctl doesn't load any data in the 
datastore but only installs the yang file. To load data (either xml or 
xml) sysrepocfg works fine.

Regards

fabio

**Comment by sankettandulwadkar** [comment]
Hi Fabio,

Here is the output -

sanket@sanket-HP-Compaq-8200-Elite-CMT-PC:~/sysrepo/build/examples$ ./application_changes_example ona100
Application will watch for changes in ona100


 ========== READING STARTUP CONFIG: ==========

/ona100:system/config(container)
/ona100:bridge/controllers (container)
/ona100:bridge/ports (container)

 ========== STARTUP CONFIG APPLIED AS RUNNING ==========

I didn't import anything. I was under the impression that the sysrepoctl --install command gets the installed yang model into the datastore. I tried importing ona100.yang and this is what I got:

sanket@sanket-HP-Compaq-8200-Elite-CMT-PC:~$ sysrepocfg --import=/home/sanket/ona100.yang --format=json ona100
[ERR] (srcfg_import_datastore:795) Unable to parse the input data: Invalid JSON data (missing top level begin-object). ()
Errors were encountered during importing. Cancelling the operation.
sanket@sanket-HP-Compaq-8200-Elite-CMT-PC:~$ sysrepocfg --import=/home/sanket/ona100.yang ona100
[ERR] (srcfg_import_datastore:795) Unable to parse the input data: Encountered invalid character sequence "module ona". ()
Errors were encountered during importing. Cancelling the operation.

I am trying to import a yang file and it gives me an invalid json error, which is not the case.

**Comment by fabioantonini** [comment]
Hi Sanket
following my last mail, you should try to import a json or xml file. Take a look at ietf-interfaces.data.xml under sysrepo/examples/yang. This is an example of a data file in xml format for ietf-interfaces.yang.
The output you see from the application shows that your repository is empty. Just the container names are displayed. You cannot import a yang file into the repository. The yang file is just the schema and it has to be installed by sysrepoctl. The data can be loaded into the datastore by sysrepocfg (--import option for example).
regards
fabio

**Comment by michalvasko** [comment]
Hi Sanket,
I will just elaborate a bit on what Fabio said and stress the important part. You should read some RFCs, mainly the [NETCONF one](https://tools.ietf.org/html/rfc6241), so that you understand the difference between **schema** (also called model or module) and **data**. The schema is simply a description and also restriction of the data that can be coupled with it. Generally, *sysrepocfg* works with data, *sysrepoctl* witch schema.

Regards,
Michal

**Comment by rkrejci** [comment]
Hi, it's not necessary to read whole RFC. There are useful tutorials to get important information:
- https://www.ietf.org/edu/technical-tutorials.html#netconfandyang
- https://github.com/mbj4668/pyang/wiki/Tutorial

**Comment by sankettandulwadkar** [comment]
Apologies. I was not even thinking about the fact that import is to import data for a specified yang. I now understand what the flow is. 

I played around with ietf-interfaces and I observed that it works fine. I was able to get, delete and set that specific example that Fabio mentions here. I also noticed that that data is "pre-loaded when i start the application_example. That means, the data is already presistent in the startup(?) datastore. Is my assumption correct?

I then added ona100.yang model to sysrepo using sysrepoctl. I could see it using sysrepoctl --list. I then created a similar .data.xml file for ona100 with this info -

<ona100 xmlns="urn:ietf:params:xml:ns:yang:ona100">
      <system>
	<configuration>
		<ona-mode>2</ona-mode>
	</configuration>
	</system>
</ona100>

I did not use pyang to do so. Do I need to always use it?

I then tried to import it.



sanket@sanket-HP-Compaq-8200-Elite-CMT-PC:~$ sysrepocfg --import=/home/sanket/ona100.data.xml --datastore=startup ona100 -l 4
[DBG] (sr_log_stderr:141) Setting log level for stderr logs to 4.
[DBG] (sr_connect:390) Connecting to Sysrepo Engine.
[DBG] (cl_socket_connect:376) Connecting to socket=/tmp/sysrepod.sock
[INF] (sr_connect:445) Connected to daemon Sysrepo Engine at socket=/tmp/sysrepod.sock
[DBG] (cl_version_verify:429) Sending version-verify request.
[DBG] (cl_version_verify:440) version-verify request sent, waiting for response.
[DBG] (cl_version_verify:452) version-verify response received, processing.
[DBG] (cl_request_process:492) Sending session-start request.
[DBG] (cl_request_process:515) session-start request sent, waiting for response.
[DBG] (cl_request_process:538) session-start response received, processing.
[DBG] (srcfg_load_module_schema:168) Loading module schema: '/home/sanket/sysrepo/build/repository/yang/ona100@2015-10-26.yang'.
[DBG] (cl_request_process:492) Sending discard-changes request.
[DBG] (cl_request_process:515) discard-changes request sent, waiting for response.
[DBG] (cl_request_process:538) discard-changes response received, processing.
[DBG] (cl_request_process:492) Sending get-items request.
[DBG] (cl_request_process:515) get-items request sent, waiting for response.
[DBG] (cl_request_process:538) get-items response received, processing.
[DBG] (cl_request_process:492) Sending get-items request.
[DBG] (cl_request_process:515) get-items request sent, waiting for response.
[DBG] (cl_request_process:538) get-items response received, processing.
[DBG] (sr_get_item_next:1105) All items has been read for xpath '/ona100:*//.'
**[DBG] (srcfg_import_datastore:891) No changes were made.**
The new configuration was successfully applied.
[DBG] (cl_request_process:492) Sending session-stop request.
[DBG] (cl_request_process:515) session-stop request sent, waiting for response.
[DBG] (cl_request_process:538) session-stop response received, processing.
sanket@sanket-HP-Compaq-8200-Elite-CMT-PC:~$ sysrepocfg --import=/home/sanket/ona100.data.xml --datastore=running ona100 -l 4
[DBG] (sr_log_stderr:141) Setting log level for stderr logs to 4.
[DBG] (sr_connect:390) Connecting to Sysrepo Engine.
[DBG] (cl_socket_connect:376) Connecting to socket=/tmp/sysrepod.sock
[INF] (sr_connect:445) Connected to daemon Sysrepo Engine at socket=/tmp/sysrepod.sock
[DBG] (cl_version_verify:429) Sending version-verify request.
[DBG] (cl_version_verify:440) version-verify request sent, waiting for response.
[DBG] (cl_version_verify:452) version-verify response received, processing.
[DBG] (cl_request_process:492) Sending session-start request.
[DBG] (cl_request_process:515) session-start request sent, waiting for response.
[DBG] (cl_request_process:538) session-start response received, processing.
[DBG] (cl_request_process:492) Sending check-enabled-running request.
[DBG] (cl_request_process:515) check-enabled-running request sent, waiting for response.
[DBG] (cl_request_process:538) check-enabled-running response received, processing.
[DBG] (srcfg_load_module_schema:168) Loading module schema: '/home/sanket/sysrepo/build/repository/yang/ona100@2015-10-26.yang'.
[DBG] (cl_request_process:492) Sending discard-changes request.
[DBG] (cl_request_process:515) discard-changes request sent, waiting for response.
[DBG] (cl_request_process:538) discard-changes response received, processing.
[DBG] (cl_request_process:492) Sending get-items request.
[DBG] (cl_request_process:515) get-items request sent, waiting for response.
[DBG] (cl_request_process:538) get-items response received, processing.
[DBG] (cl_request_process:492) Sending get-items request.
[DBG] (cl_request_process:515) get-items request sent, waiting for response.
[DBG] (cl_request_process:538) get-items response received, processing.
[DBG] (sr_get_item_next:1105) All items has been read for xpath '/ona100:*//.'
**[DBG] (srcfg_import_datastore:891) No changes were made.**
The new configuration was successfully applied.
[DBG] (cl_request_process:492) Sending session-stop request.
[DBG] (cl_request_process:515) session-stop request sent, waiting for response.
[DBG] (cl_request_process:538) session-stop response received, processing.


I always get "no changes were made" message. If sysrepocfg reads the ona100 yang, i believe it knows of the module. What is it that I am missing?

--Sanket

**Comment by fabioantonini** [comment]
Hi Sanket

about pyang you don't need it to work with sysrepo. Nevertheless pyang 
is a valuable tool to verify/display yang files. On my side we use pyang 
to validate new yang files or modified one.

About the sentence "*[DBG] (srcfg_import_datastore:891) No changes were 
made."*

this occurs because probably you are trying to import the same xml file 
to the datastore for a second time. sysrepocfg verifies the differences 
between the data currently available in the datastore a the xml files. 
It sounds as there is no difference. Please try to read also the content 
of the datastore with sysrepocfg --export .......just to verify the data 
available.

In any case it seems to be working fine

regards

fabio

**Comment by fabioantonini** [comment]
Hi Sanket
Did you try pyang? is behavior more clear now? 
regards

fabio
